### ISSUE TITLE
variable kerning doesn't seem to activate

### ISSUE THREAD

**Issue by rsheeter** [googlefonts/fontc#402]
After https://github.com/googlefonts/fontc/pull/373 I seem to be seeing that variable kerning doesn't actually activate. For example:

```shell
$ export RUST_LOG=info,fontbe=trace
$ rm -rf build/ && cargo run -p fontc -- --source resources/testdata/glyphs3/WghtVar.glyphs
...noise
# fontc generated kerning

@public.kern1.bracketright_R = [bracketright];
@public.kern1.bracketleft_R = [bracketleft];
@public.kern2.bracketright_L = [bracketright];
@public.kern2.bracketleft_L = [bracketleft];


feature kern {
  pos exclam @public.kern2.bracketright_L (wght=400:-160 wght=700:0);
  pos exclam exclam (wght=400:-360 wght=700:-100);
  pos hyphen hyphen (wght=400:-150 wght=700:-50);
  pos bracketleft bracketright (wght=400:-300 wght=700:-150);
  pos @public.kern1.bracketleft_R exclam (wght=400:-165 wght=700:0);
  pos exclam hyphen (wght=400:20 wght=700:0);
} kern;
...

# OK, great, [] should kern -300 at wght=400, -150 at wght=700 per pos bracketleft bracketright (wght=400:-300 wght=700:-150);
# [] have width=600
$  ../harfbuzz/build/util/hb-shape build/font.ttf --text='[' --no-clusters
[bracketleft+600]
$  ../harfbuzz/build/util/hb-shape build/font.ttf --text=']' --no-clusters
[bracketright+600]

# 400 seems to work, bracketleft isn't 600 wide
$ ../harfbuzz/build/util/hb-shape build/font.ttf --text='[]' --no-clusters
[bracketleft+300|bracketright+600]

# but if we push toward 700 bracketleft doesn't change width, it just sits at 300
$ ../harfbuzz/build/util/hb-shape build/font.ttf --text='[]' --no-clusters --variations="wght=500"
[bracketleft+300|bracketright+600]
$ ../harfbuzz/build/util/hb-shape build/font.ttf --text='[]' --no-clusters --variations="wght=700"
[bracketleft+300|bracketright+600]
```

**Comment by cmyr** [comment]
So there are a few weird things going on here, but a major one is that we aren't building the GDEF table, and that's where the variation data is stored.

**Comment by anthrotype** [comment]
so are we just throwing away the ItemVariationStore that write-fonts built?

**Comment by anthrotype** [comment]
somewhere here we should store the GDEF we get from fea-rs, like we do for GPOS/GSUB:
https://github.com/googlefonts/fontc/blob/a4ee74ce44df64ccdcd76476f8cc3076d9b0b4c6/fontbe/src/features.rs#L147-L165

**Comment by anthrotype** [comment]
With #405 merged, we now have a GDEF with an ItemVariationStore that is doing something, even though not exactly correct yet:

```
# the advance width of bracketleft should be 450 (600 - 150) at wght=700, but we get 150...
$ hb-shape build/font.ttf --text='[]' --no-clusters --variations="wght=700"
[bracketleft+150|bracketright+600]
```

**Comment by anthrotype** [comment]
the GDEF table that we get from fea-rs for the resources/testdata/glyphs3/WghtVar.glyphs looks like this:

```xml
  <GDEF>
    <Version value="0x00010003"/>
    <VarStore Format="1">
      <Format value="1"/>
      <VarRegionList>
        <!-- RegionAxisCount=1 -->
        <!-- RegionCount=2 -->
        <Region index="0">
          <VarRegionAxis index="0">
            <StartCoord value="0.0"/>
            <PeakCoord value="0.0"/>
            <EndCoord value="0.0"/>
          </VarRegionAxis>
        </Region>
        <Region index="1">
          <VarRegionAxis index="0">
            <StartCoord value="0.0"/>
            <PeakCoord value="1.0"/>
            <EndCoord value="1.0"/>
          </VarRegionAxis>
        </Region>
      </VarRegionList>
      <!-- VarDataCount=1 -->
      <VarData index="0">
        <!-- ItemCount=6 -->
        <NumShorts value="2"/>
        <!-- VarRegionCount=2 -->
        <VarRegionIndex index="0" value="0"/>
        <VarRegionIndex index="1" value="1"/>
        <Item index="0" value="[20, -20]"/>
        <Item index="1" value="[-150, 100]"/>
        <Item index="2" value="[-300, 150]"/>
        <Item index="3" value="[-160, 160]"/>
        <Item index="4" value="[-165, 165]"/>
        <Item index="5" value="[-360, 260]"/>
      </VarData>
    </VarStore>
  </GDEF>
```

There's an extra region at index=0 of VarRegionList, with wght=(min=0, peak=0, max=0) which looks like it should not be there, not sure yet how it ended up there

**Comment by anthrotype** [comment]
found the culprit and fixed in [632db7a](https://github.com/googlefonts/fontc/pull/373/commits/632db7a0a289330b426a18c3373c80c2f0610f11)

Finally we have variable kerning! 🎉

**Comment by anthrotype** [comment]
> Finally we have variable kerning!

ok well.. at least we do for the simple WghtVar.glyphs test font...

I tested Oswald as well, comparing it with the fontmake-built font, and I am still seeing some pairs that do not get any kerning at all. But this happens regardless of variations, even at the default instance I am seeing differences.

E.g. "Va" kerning is working but "aV" is not, but it should.. 🤔

```
$ hb-shape build/Oswald-VF-fontc.ttf --text='VaV' --no-clusters --features=-kern
[V+505|a+426|V+505]
$ hb-shape build/Oswald-VF-fontmake.ttf --text='VaV' --no-clusters --features=-kern
[V+505|a+426|V+505]
$ hb-shape build/Oswald-VF-fontc.ttf --text='VaV' --no-clusters
[V+455|a+426|V+505]
$ hb-shape build/Oswald-VF-fontmake.ttf --text='VaV' --no-clusters
[V+455|a+378|V+505]
```
 
I'm suspecting it could have something to do with PairPos subtables being split, cf. https://github.com/googlefonts/fontations/pull/401/ (though I haven't been able to confirm this so far)

**Comment by anthrotype** [comment]
ok, I figured the reason was for class to glyph or glyph to class kerning pairs (also known as class kerning "exceptions") such as the offending `@a V` pair in Oswald, we were also using `pos` instead of `enum pos`, so instead of expanding them to specific glyph-glyph pairs, the compiler was using class pairs for these, which leads to conflicting/invalid class memberships. ufo2ft kernFeatureWriter adds `enum pos` to all `enumerated = pair.firstIsClass ^ pair.secondIsClass`, see https://github.com/googlefonts/ufo2ft/blob/b3895a96ca910c1764df016bfee4719448cfec4a/Lib/ufo2ft/featureWriters/kernFeatureWriter.py#L360C26-L360C39

I have fixed it in https://github.com/googlefonts/fontc/pull/373/commits/9661684f112dfdd9c4554a2cdf8dd0e5f98ecbde and now I am seeing variable kerning work for Oswald. I have only done a spot visual check, not a deep diff, but results are encouraging.
